ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+        (bpf_attach_type_t)_ebpf_netevent_event_section_info->bpf_attach_type;
+    _netevent_ebpf_netevent_event_hook_provider_data.link_type = BPF_LINK_TYPE_PLAIN;
+    status = ebpf_extension_hook_provider_register(
+        &hook_provider_parameters,
+        _netevent_ebpf_extension_netevent_event_on_client_attach,
+        _netevent_ebpf_extension_netevent_event_on_client_detach,
+        NULL,
+        &_ebpf_netevent_event_hook_provider_context);
+    if (status != EBPF_SUCCESS) {
+        EBPF_EXT_LOG_MESSAGE_NTSTATUS(
+            EBPF_EXT_TRACELOG_LEVEL_ERROR,
+            EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+            "ebpf_extension_hook_provider_register",
+            status);
+        goto Exit;
+    }
+
+    // Initialize the global event buffer.
+    _event_buffer = (uint8_t*)ExAllocatePoolUninitialized(NonPagedPoolNx, _event_buffer_size, EBPF_EXTENSION_POOL_TAG);
+    if (_event_buffer == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        EBPF_EXT_LOG_MESSAGE_NTSTATUS(
+            EBPF_EXT_TRACELOG_LEVEL_ERROR,
+            EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+            "Insufficient memory initializing the event buffer",
+            status);
+        goto Exit;
+    }
+
+Exit:
+    if (!NT_SUCCESS(status)) {
+        ebpf_ext_unregister_netevent();
+    }
+    EBPF_EXT_RETURN_NTSTATUS(status);
+}
+
+void
+ebpf_ext_unregister_netevent()
+{
+    if (_ebpf_netevent_event_hook_provider_context) {
+        ebpf_extension_hook_provider_unregister(_ebpf_netevent_event_hook_provider_context);
+        _ebpf_netevent_event_hook_provider_context = NULL;
+    }
+    if (_ebpf_netevent_event_program_info_provider_context) {
+        ebpf_extension_program_info_provider_unregister(_ebpf_netevent_event_program_info_provider_context);
+        _ebpf_netevent_event_program_info_provider_context = NULL;
+    }
+    if (_event_buffer) {
+        ExFreePool(_event_buffer);
+        _event_buffer = NULL;
+    }
+}
+
+//
+// eBPF NetEvent Program Information NPI helper routines.
+//
+static ebpf_result_t
+_ebpf_netevent_program_context_create(
+    _In_reads_bytes_opt_(data_size_in) const uint8_t* data_in,
+    size_t data_size_in,
+    _In_reads_bytes_opt_(context_size_in) const uint8_t* context_in,
+    size_t context_size_in,
+    _Outptr_ void** context)
+{
+    EBPF_EXT_LOG_ENTRY();
+    ebpf_result_t result;
+    netevent_event_md_t* netevent_event_context = NULL;
+
+    *context = NULL;
+
+    if (context_in == NULL || context_size_in < sizeof(netevent_event_md_t)) {
+        EBPF_EXT_LOG_MESSAGE(EBPF_EXT_TRACELOG_LEVEL_ERROR, EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, "Context is required");
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    // Allocate memory for the context.
+    netevent_event_context = (netevent_event_md_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(netevent_event_md_t), EBPF_EXTENSION_POOL_TAG);
```

Comment:
Shouldn't the pool tag be unique for each extension?

Reply:
After discussing offline, this does need to be private to each extension, and the library needs to be modified to receive the pool tag as a parameter.
Will open an issue to track this.

Reply:
https://github.com/microsoft/ntosebpfext/issues/61